### ISSUE TITLE
nix-flatpak keeps on uninstalling a remote before adding the remote from a flatpakref

### ISSUE THREAD

**Issue by SpiderUnderUrBed** [gmodena/nix-flatpak#90]
What I stated on the title is what I am assuming to happen given that until the recent pull request, I got a diffrent kind of error on how it was unable to remove the remote. I realize that I need to figure out why nix-flatpak wants to remove the remote from my flatpakref and I cannot figure it out. I stated I had a issue like this earlier, here is the error:
```
× flatpak-managed-install.service
     Loaded: loaded (/etc/systemd/system/flatpak-managed-install.service; enabled; preset: ignored)
     Active: failed (Result: exit-code) since Tue 2024-10-22 07:06:22 NZDT; 286ms ago
 Invocation: 63a7fa1b138344b389ebca6c11d44305
    Process: 111503 ExecStart=/nix/store/3w9yrhksarir324f0r01nvg349x7xlvg-flatpak-managed-install (code=exited, status=1/FAILURE)
   Main PID: 111503 (code=exited, status=1/FAILURE)
         IP: 1.5M in, 25.8K out
   Mem peak: 37.7M
        CPU: 899ms

Oct 22 07:06:11 daspidercave systemd[1]: Starting flatpak-managed-install.service...
Oct 22 07:06:16 daspidercave flatpak[112474]: system: Removed remote sober
Oct 22 07:06:18 daspidercave 3w9yrhksarir324f0r01nvg349x7xlvg-flatpak-managed-install[112478]: Skipping: moe.launcher.the-honkers-railway-launcher/x86_64/master is already installed
Oct 22 07:06:20 daspidercave 3w9yrhksarir324f0r01nvg349x7xlvg-flatpak-managed-install[112507]: Skipping: moe.launcher.an-anime-game-launcher/x86_64/master is already installed
Oct 22 07:06:22 daspidercave 3w9yrhksarir324f0r01nvg349x7xlvg-flatpak-managed-install[112519]: error: No remote refs found for ‘org.vinegarhq.Sober’
Oct 22 07:06:22 daspidercave systemd[1]: flatpak-managed-install.service: Main process exited, code=exited, status=1/FAILURE
Oct 22 07:06:22 daspidercave systemd[1]: flatpak-managed-install.service: Failed with result 'exit-code'.
Oct 22 07:06:22 daspidercave systemd[1]: Failed to start flatpak-managed-install.service.
Oct 22 07:06:22 daspidercave systemd[1]: flatpak-managed-install.service: Consumed 899ms CPU time, 37.7M memory peak, 1.5M incoming IP traffic, 25.7K outgoing IP traffic.
warning: error(s) occurred while switching to the new configuration
```
And here is the flatpak.nix file:
https://pastebin.com/XMY5iwmu

Is this expected behavior from uninstallUnmanaged when a remote suggested from a flatpakref file isnt also listed explictly under "remotes"?

**Comment by gmodena** [comment]
hey @SpiderUnderUrBed 

> Is this expected behavior from uninstallUnmanaged when a remote suggested from a flatpakref file isnt also listed explictly under "remotes"?

What you experience is not expected behaviour, and unfortunately I am unable to reproduce it in my test envs.

Could you maybe share some more verbose systemd/journald logs, and the whole content of `/nix/store/3w9yrhksarir324f0r01nvg349x7xlvg-flatpak-managed-install` ?

As a last resort, could you try starting from a fresh system? Meaning: manually removing all unmanaged remotes and deleting `/nix/var/nix/gcroots/flatpak-state.json`?

**Comment by SpiderUnderUrBed** [comment]
https://pastebin.com/hqDp4Cda
^ content of /nix/store/3w9yrhksarir324f0r01nvg349x7xlvg-flatpak-managed-install
https://pastebin.com/BPs4zLhk
^ journalctl logs going back 500.

This is what happens which I tried what you said:
https://pastebin.com/RDMJ6Ywe

Also uninstalled remotes should have been removed automatically, and to my knowlage they have, given this happens with my flatpakref install, the error must lie with that.

**Comment by gmodena** [comment]
> This is what happens which I tried what you said:
https://pastebin.com/RDMJ6Ywe

You removed nix-flatpak's state file, but I'm afraid you may still have 'org.vinegarhq.Sober' installed from previous installations (either manually or via nix-flatpak).

> Oct 30 09:09:48 daspidercave f3szjfavynaafcvixr7jxyyhqcf50a3c-flatpak-managed-install[12356]: error: No remote refs found for ‘org.vinegarhq.Sober’

You are seeing this message because nix-flatpak found 'org.vinegarhq.Sober' already present in your system (using `flatpak list --app`) and is attempting to reinstall it from the URI rather than the provided flatpakref file (and thus also installing the remote).

I was able to repro under these conditions.

Could you:
- manually uninstall `org.vinegarhq.Sober`.
- manually uninstall any `sober` remote present in your system.
- remove nix-flatpak's state file.
- run another activation.

And just to be sure: point nix-flatpak to https://sober.vinegarhq.org/sober.flatpakref, instead of a local file (unlikely to be an issue, unless you modified `sober.flatpakref`).

**Comment by SpiderUnderUrBed** [comment]
I will test all of that later, I suspect another issue is happening when it comes to flatpak refs updating, I have to manually uninstall the flatpak then re-install the flatpak ref when i want to update, any idea on why this is?

**Comment by gmodena** [comment]
I'm testing a fix. I should have patch incoming sometimes this week.

**Comment by gmodena** [comment]
#93 should fix the issue (and others).

The patch can be tested by tracking the `fix-flatpakref-install` branch.

**Comment by gmodena** [comment]
Reposting this comment from #93 for visibility:

I'm really at a loss here. Let me walk you through what I did  to test #93 (now in `main`). 

Note: the snippets below are from a VM, but the same behavior occurs on my daily driver, which has a much more complex configuration. Can you help me pinpoint what might be different on my end, compared to what you are doing?

## Setup

I'm loading nix flatpak as a nixos module. The flake input is set to:
```
flatpaks.url = "github:gmodena/nix-flatpak?ref=fix-flatpakref-install";
```

## Flatpakref in a freshly built VM
I spin up a freshly built vm (build atop the config in `testing-base`). This is what the initial flatpak config looks like:
```
# flatpak.nix
{ lib, ... }: {

  services.flatpak.remotes = lib.mkOptionDefault [{
    name = "flathub-beta";
    location = "https://flathub.org/beta-repo/flathub-beta.flatpakrepo";
  }];

  services.flatpak.update.auto.enable = true;
  services.flatpak.uninstallUnmanaged = true;
  services.flatpak.packages = [
    { 
      flatpakref="https://sober.vinegarhq.org/sober.flatpakref";
      sha256="1pj8y1xhiwgbnhrr3yr3ybpfis9slrl73i0b1lc9q89vhip6ym2l";
    }
  ];
}
```

At boot, Sober and its remote are installed:
```
[root@nixos:~]# flatpak remotes
Name         Options
flathub      system
flathub-beta system
sober        system

[root@nixos:~]# flatpak list --app
Name        Application ID            Version             Branch       Installation
Sober       org.vinegarhq.Sober       0.0.0-78a4e30       master       system

[root@nixos:~]# cat /nix/var/nix/gcroots/flatpak-state.json 
{"overrides":{},"packages":["org.vinegarhq.Sober"],"remotes":["flathub","flathub-beta","sober"]}
```

I build the first generation with:
```
[root@nixos:~/config]# nixos-rebuild switch --flake '.'
```

Which is
```
[root@nixos:~/config]# nixos-rebuild list-generations
Generation  Build-date           NixOS version           Kernel  Configuration Revision  Specialisation
1 current   2024-11-11 22:17:47  23.11.20240709.205fd42  6.1.96                          *
```

Sober's remote and flatpak-state are the same as the build. Now I'll run `nixos-rebuild switch --flake '.'` a couple more times.

I tried adding new Flatpaks at activations 3 (`im.riot.Riot`)  and 4 (`com.logseq.Logseq`), and removed them at 5 and 6 to see if that would trigger any unusual code paths. Every activation succeeded. 

```
[root@nixos:~/config]# nixos-rebuild list-generations
Generation  Build-date           NixOS version           Kernel  Configuration Revision  Specialisation
7 current   2024-11-11 22:36:16  23.11.20240709.205fd42  6.1.96                          *
6           2024-11-11 22:35:38  23.11.20240709.205fd42  6.1.96                          *
5           2024-11-11 22:27:49  23.11.20240709.205fd42  6.1.96                          *
4           2024-11-11 22:23:14  23.11.20240709.205fd42  6.1.96                          *
3           2024-11-11 22:22:18  23.11.20240709.205fd42  6.1.96                          *
2           2024-11-11 22:20:26  23.11.20240709.205fd42  6.1.96                          *
1           2024-11-11 22:17:47  23.11.20240709.205fd42  6.1.96                          *
```

By generation 7, I only have Sober.
```
[root@nixos:~/config]# flatpak remotes
Name         Options
flathub      system
flathub-beta system
sober        system

[root@nixos:~/config]# flatpak list --app
Name        Application ID            Version             Branch       Installation
Sober       org.vinegarhq.Sober       0.0.0-78a4e30       master       system
```

Let's check the service status:
```
○ flatpak-managed-install.service
     Loaded: loaded (/etc/systemd/system/flatpak-managed-install.service; enabled; preset: enabled)
     Active: inactive (dead) since Mon 2024-11-11 22:36:19 UTC; 1min 22s ago
TriggeredBy: ● flatpak-managed-install.timer
    Process: 11137 ExecStart=/nix/store/1cazmcxv9nb97il5fxai0mixp5xk1gng-flatpak-managed-install (code=exited, status=0/SUCCESS)
   Main PID: 11137 (code=exited, status=0/SUCCESS)
         IP: 1.4M in, 15.2K out
        CPU: 691ms

Nov 11 22:36:17 nixos systemd[1]: Starting flatpak-managed-install.service...
Nov 11 22:36:18 nixos 1cazmcxv9nb97il5fxai0mixp5xk1gng-flatpak-managed-install[11247]:  1.                   com.logseq.Logseq        stable        r
Nov 11 22:36:18 nixos 1cazmcxv9nb97il5fxai0mixp5xk1gng-flatpak-managed-install[11247]: Uninstalling…
Nov 11 22:36:18 nixos flatpak[11247]: system: Uninstalled app/com.logseq.Logseq/x86_64/stable
Nov 11 22:36:18 nixos 1cazmcxv9nb97il5fxai0mixp5xk1gng-flatpak-managed-install[11247]: Uninstall complete.
Nov 11 22:36:19 nixos systemd[1]: flatpak-managed-install.service: Deactivated successfully.
Nov 11 22:36:19 nixos systemd[1]: Finished flatpak-managed-install.service.
Nov 11 22:36:19 nixos systemd[1]: flatpak-managed-install.service: Consumed 691ms CPU time, received 1.4M IP traffic, sent 15.2K IP traffic.
```

## Remove and then re-install a flatpakref

As a further test, I removed sober (by commenting out the flatpakref config) and  then re-installed it.
```
  services.flatpak.packages = [
  #  { 
  #    flatpakref="https://sober.vinegarhq.org/sober.flatpakref";
  #    sha256="1pj8y1xhiwgbnhrr3yr3ybpfis9slrl73i0b1lc9q89vhip6ym2l";
  #  }
  ];
```

```
[root@nixos:~/config]# nixos-rebuild switch --flake '.'

Generation  Build-date           NixOS version           Kernel  Configuration Revision  Specialisation
8 current   2024-11-11 22:42:45  23.11.20240709.205fd42  6.1.96                          *
7           2024-11-11 22:36:16  23.11.20240709.205fd42  6.1.96                          *
6           2024-11-11 22:35:38  23.11.20240709.205fd42  6.1.96                          *
5           2024-11-11 22:27:49  23.11.20240709.205fd42  6.1.96                          *
4           2024-11-11 22:23:14  23.11.20240709.205fd42  6.1.96                          *
3           2024-11-11 22:22:18  23.11.20240709.205fd42  6.1.96                          *
2           2024-11-11 22:20:26  23.11.20240709.205fd42  6.1.96                          *
1           2024-11-11 22:17:47  23.11.20240709.205fd42  6.1.96                          *

```

By generation 8 there is no Sober installed:
```
[root@nixos:~/config]# flatpak remotes
Name         Options
flathub      system
flathub-beta system

[root@nixos:~/config]# flatpak list --app


[root@nixos:~/config]# cat /nix/var/nix/gcroots/flatpak-state.json 
{"overrides":{},"packages":[],"remotes":["flathub","flathub-beta"]}
```

Let's check the service status:
```
[root@nixos:~/config]# systemctl status flatpak-managed-install
○ flatpak-managed-install.service
     Loaded: loaded (/etc/systemd/system/flatpak-managed-install.service; enabled; preset: enabled)
     Active: inactive (dead) since Mon 2024-11-11 22:42:47 UTC; 2min 14s ago
TriggeredBy: ● flatpak-managed-install.timer
    Process: 11796 ExecStart=/nix/store/kn9q9kgxfimyb40pl99bmmabanh9fspp-flatpak-managed-install (code=exited, status=0/SUCCESS)
   Main PID: 11796 (code=exited, status=0/SUCCESS)
         IP: 21.4K in, 3.8K out
        CPU: 367ms

Nov 11 22:42:46 nixos systemd[1]: Starting flatpak-managed-install.service...
Nov 11 22:42:46 nixos kn9q9kgxfimyb40pl99bmmabanh9fspp-flatpak-managed-install[11906]:  1.                   org.vinegarhq.Sober        master        r
Nov 11 22:42:46 nixos kn9q9kgxfimyb40pl99bmmabanh9fspp-flatpak-managed-install[11906]: Uninstalling…
Nov 11 22:42:46 nixos flatpak[11906]: system: Uninstalled app/org.vinegarhq.Sober/x86_64/master
Nov 11 22:42:47 nixos kn9q9kgxfimyb40pl99bmmabanh9fspp-flatpak-managed-install[11906]: Uninstall complete.
Nov 11 22:42:47 nixos flatpak[11931]: system: Removed remote sober
Nov 11 22:42:47 nixos systemd[1]: flatpak-managed-install.service: Deactivated successfully.
Nov 11 22:42:47 nixos systemd[1]: Finished flatpak-managed-install.service.
Nov 11 22:42:47 nixos systemd[1]: flatpak-managed-install.service: Consumed 367ms CPU time, received 21.4K IP traffic, sent 3.8K IP traffic.

```

Now let's re-install Sober:
```
  services.flatpak.packages = [
    { 
      flatpakref="https://sober.vinegarhq.org/sober.flatpakref";
      sha256="1pj8y1xhiwgbnhrr3yr3ybpfis9slrl73i0b1lc9q89vhip6ym2l";
    }
  ];
```
After activation:
```
Generation  Build-date           NixOS version           Kernel  Configuration Revision  Specialisation
9 current   2024-11-11 22:45:45  23.11.20240709.205fd42  6.1.96                          *
8           2024-11-11 22:42:45  23.11.20240709.205fd42  6.1.96                          *
7           2024-11-11 22:36:16  23.11.20240709.205fd42  6.1.96                          *
6           2024-11-11 22:35:38  23.11.20240709.205fd42  6.1.96                          *
5           2024-11-11 22:27:49  23.11.20240709.205fd42  6.1.96                          *
4           2024-11-11 22:23:14  23.11.20240709.205fd42  6.1.96                          *
3           2024-11-11 22:22:18  23.11.20240709.205fd42  6.1.96                          *
2           2024-11-11 22:20:26  23.11.20240709.205fd42  6.1.96                          *
1           2024-11-11 22:17:47  23.11.20240709.205fd42  6.1.96                          *

```

At generation 9, sober has been installed again:
```
[root@nixos:~/config]# flatpak remotes
Name         Options
flathub      system
flathub-beta system
sober        system

[root@nixos:~/config]# flatpak list --app
Name        Application ID            Version             Branch       Installation
Sober       org.vinegarhq.Sober       0.0.0-78a4e30       master       system

[root@nixos:~/config]# cat /nix/var/nix/gcroots/flatpak-state.json 
{"overrides":{},"packages":["org.vinegarhq.Sober"],"remotes":["flathub","flathub-beta","sober"]}
```

I've run nixos-rebuild a couple more times. Let's check the service status:
```
[root@nixos:~/config]# systemctl status flatpak-managed-install
○ flatpak-managed-install.service
     Loaded: loaded (/etc/systemd/system/flatpak-managed-install.service; enabled; preset: enabled)
     Active: inactive (dead) since Mon 2024-11-11 22:49:18 UTC; 1s ago
TriggeredBy: ● flatpak-managed-install.timer
    Process: 13354 ExecStart=/nix/store/1cazmcxv9nb97il5fxai0mixp5xk1gng-flatpak-managed-install (code=exited, status=0/SUCCESS)
   Main PID: 13354 (code=exited, status=0/SUCCESS)
         IP: 70.4K in, 9.4K out
        CPU: 328ms

Nov 11 22:49:17 nixos systemd[1]: Starting flatpak-managed-install.service...
Nov 11 22:49:18 nixos systemd[1]: flatpak-managed-install.service: Deactivated successfully.
Nov 11 22:49:18 nixos systemd[1]: Finished flatpak-managed-install.service.
Nov 11 22:49:18 nixos systemd[1]: flatpak-managed-install.service: Consumed 328ms CPU time, received 70.4K IP traffic, sent 9.4K IP traffic.
```

**Comment by SpiderUnderUrBed** [comment]
The issue is no longer present in my system. When testing the fix, I specified the rev before commit e637f51f6523c8fad28af79ea6b1f64697fa9945 (as that commit was made much later, mid-discussion). Now after multiple tries under different conditions I no longer have the issue on my end, so I think that commit resolved my issue. However to ensure that this is not a false-flag, I want to hear if @Bear-03 no longer has the issue before marking this issue as solved. Also I tested it and it definitely seems that the latest commit in that branch solved my issue.
